docs/type_hints.rst

```diff
@@ -0,0 +1,100 @@
+
+Intended Type Hints
+===================
+
+The intent is to have type hints in place for all of RDFLib and to have these type hints as accurate and correct as possible.
+
+The correctness and accuracy of type hints are determined by both the standards that RDFLib aims to conform to, like RDF 1.1, and the deliberate choices that are made when implementing RDFLib. For example, given that the RDF 1.1 specification stipulates that the subject of an RDF triple cannot be a literal, all functions that accept an *RDF term* to be used as the subject of a triple should have type hints which excludes values that are literals.
+
+There may be cases where some functionality of RDFLib may work perfectly well with values of types that are excluded by the type hints, but if these additional types violate the relevant standards we will consider the correct type hints to be those that exclude values of these types.
+
+Public Type Aliases
+=============================
+In python, type hints are specified in annotations. Type hints are different from type aliases which are normal python variables that are not intended to provide runtime utility and are instead intended for use in static type checking.
+
+For clarify, the following is an example of a function ``foo`` with type hints:
+
+.. code-block:: python
+	
+    def foo(a: int) -> int:
+        return a + 1
+
+While the following is an example of a type alias ``Bar```:
+
+.. code-block:: python
+
+    from typing import Tuple
+
+    Bar = Tuple[int, str]
+
+RDFLib will provide public type aliases under the ``rdflib.typing`` package, for example, ``rdflib.typing.Triple``, ``rdflib.typing.Quad``. Type aliases in the rest of RDFLib should be private (i.e. being with an underscore).
+
+Versioning, Compatibility and Stability
+=======================================
+
+RDFLib attempts to adhere to `semver 2.0 <https://semver.org/spec/v2.0.0.html>`_ which is concerned with the public API of software.
+
+Ignoring type hints, the public API of RDFLib exists implicitly as a consequence of the code of RDFLib and the actual behaviour this entails, the relevant standards that RDFLib is trying to implement, and the documentation of RDFLib, with some interplay between all three of these. RDFLib's public API includes public type aliases, as these are normal python variables and not annotations.
+
+Type hints attempt to formally document RDFLib's implicitly defined public API in a machine-readable fashion as accurately and correctly as possible within the framework outline earlier in this document.
+
+Type hints do not change the functionality of RDFLib, nor do they affect the runtime API of RDFLib. In this way then, they are somewhat outside of the scope of semver, however, they still have an impact on the users of RDFLib, even if this impact is not at runtime, but during development. This necessitates some clarity as to what users of RDFLib should expect regarding type hints in RDFLib releases.
```

Comment:
> Type hints do not change the functionality of RDFLib, nor do they affect the runtime API of RDFLib.

This is a bit debatable.  If I type-review code with a bad signature, it failing the type review to me is a failure of functionality, though not necessarily runtime functionality.

May this instead be phrased as:
> Type hints do not affect the runtime API of RDFLib.

Reply:
I will go with: "runtime API or functionality of RDFLib"

Reply:
Changed it now.